### PR TITLE
Keyspace wrapper to check keyspace is allowed

### PR DESCRIPTION
**Pull request by fppt**

### CODE REVIEW

File: grakn-engine/src/main/java/ai/grakn/engine/factory/EngineGraknTxFactory.java [hunk 1]
```diff
@@ -67,7 +68,11 @@ public synchronized void refreshConnections(){
     }
 
     public GraknTx tx(String keyspace, GraknTxType type){
-        if(!keyspace.equals(SystemKeyspace.SYSTEM_KB_NAME)) {
+        return tx(Keyspace.of(keyspace), type);
+    }
+
+    public GraknTx tx(Keyspace keyspace, GraknTxType type){
+        if(!keyspace.getValue().equals(SystemKeyspace.SYSTEM_KB_KEYSPACE)) {
```
Comment (aelred) [inline grakn-engine/src/main/java/ai/grakn/engine/factory/EngineGraknTxFactory.java]: We should compare the keyspaces directly here, not their names.

File: grakn-graql/src/test/java/ai/grakn/graql/internal/hal/HALBuilderTest.java [hunk 2]
```diff
@@ -147,7 +148,7 @@ private Json getHALRepresentation(GraknTx graph, String queryString) {
 
     private Json getHALExploreRepresentation(GraknTx graph, String conceptId) {
         Concept concept = graph.getConcept(ConceptId.of(conceptId));
-        return Json.read(HALExploreConcept(concept, graph.getKeyspace(), 0, 5));
+        return Json.read(HALExploreConcept(concept, graph.getKeyspace().getValue(), 0, 5));
```
Comment (aelred) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/hal/HALBuilderTest.java]: `HALExploreConcept` should take `Keyspace`, not `String` I guess

File: grakn-test/test-integration/src/test/java/ai/grakn/client/BatchMutatorClientTest.java [hunk 3]
```diff
@@ -176,7 +178,7 @@ private BatchMutatorClient loader(){
             nameTag.attribute(nameTagId);
             graph.admin().commitNoLogs();
 
-            return spy(new BatchMutatorClient(graph.getKeyspace(), engine.uri(), true));
+            return spy(new BatchMutatorClient(graph.getKeyspace().getValue(), engine.uri(), true));
```
Comment (aelred) [inline grakn-test/test-integration/src/test/java/ai/grakn/client/BatchMutatorClientTest.java]: The `BatchMutatorClient` should also take a `Keyspace`

File: grakn-test/test-integration/src/test/java/ai/grakn/test/migration/csv/CSVMigratorTest.java [hunk 4]
```diff
@@ -63,9 +63,9 @@
 
     @Before
     public void setup() {
-        String keyspace = SampleKBLoader.randomKeyspace();
+        Keyspace keyspace = SampleKBLoader.randomKeyspace();
         factory = Grakn.session(engine.uri(), keyspace);
-        migrator = Migrator.to(engine.uri(), keyspace);
+        migrator = Migrator.to(engine.uri(), keyspace.getValue());
```
Comment (aelred) [inline grakn-test/test-integration/src/test/java/ai/grakn/test/migration/csv/CSVMigratorTest.java]: This should also pass the keyspace directly.

File: grakn-engine/src/main/java/ai/grakn/engine/SystemKeyspace.java [hunk 5]
```diff
@@ -69,13 +70,13 @@
...
     // If there is a more natural home for this constant, feel free to put it there! (Boris)
-    public static final String SYSTEM_KB_NAME = "graknSystem";
+    public static final Keyspace SYSTEM_KB_KEYSPACE = Keyspace.of("graknsystem");
     private static final String SYSTEM_VERSION = "system-version";
     public static final Label KEYSPACE_ENTITY = Label.of("keyspace");
     public static final Label KEYSPACE_RESOURCE = Label.of("keyspace-name");
 
     private static final Logger LOG = LoggerFactory.getLogger(SystemKeyspace.class);
-    private final ConcurrentHashMap<String, Boolean> openSpaces;
+    private final ConcurrentHashMap<Keyspace, Boolean> openSpaces;
```
[the remark this answers is not in the document]
Comment (fppt) [reply grakn-engine/src/main/java/ai/grakn/engine/SystemKeyspace.java]: lols

File: grakn-engine/src/main/java/ai/grakn/engine/controller/CommitLogController.java [hunk 6]
```diff
@@ -81,7 +78,7 @@ private String deleteConcepts(Request req, Response res){
         @ApiImplicitParam(name = COMMIT_LOG_COUNTING, value = "A Json Array types with new and removed instances", required = true, dataType = "string", paramType = "body")
     })
     private String submitConcepts(Request req, Response res) {
-        Keyspace keyspace = Keyspace.of(Optional.ofNullable(req.queryParams(KEYSPACE_PARAM)).orElse(defaultKeyspace));
+        Keyspace keyspace = Keyspace.of(req.queryParams(KEYSPACE_PARAM));
```
Comment (aelred) [inline grakn-engine/src/main/java/ai/grakn/engine/controller/CommitLogController.java]: I think there's a method `mandatoryQueryParam` that will throw if it's not there - which is what we need here.

**Comment by fppt** [comment]
@aelred re-review please!!!

**Pull request merged**